tianshou/data/batch.py

```diff
@@ -593,23 +601,69 @@ def stack_(self,
                 if not issubclass(v.dtype.type, (np.bool_, np.number)):
                     v = v.astype(np.object)
                 self.__dict__[k] = v
-        keys_partial = reduce(set.symmetric_difference, keys_map)
+        keys_partial = set.union(*keys_map) - keys_shared
         _assert_type_keys(keys_partial)
         for k in keys_partial:
+            is_dict = False
+            value = None
             for i, e in enumerate(batches):
                 val = e.get(k, None)
                 if val is not None:
-                    try:
-                        self.__dict__[k][i] = val
-                    except KeyError:
-                        self.__dict__[k] = \
-                            _create_value(val, len(batches))
-                        self.__dict__[k][i] = val
+                    if isinstance(val, (dict, Batch)):
```

Comment:
Why replacing something working and efficient by such a complicated nested if/else code ?

Reply:
Because the current code of stack_ method has different behavior when dealing with inconsistent keys:
```python
In [2]: a = Batch(a=[1], b=[2], c=[3]) 
   ...: b = Batch(a=[4], b=[5], d=[6]) 
   ...: c = Batch(c=[3], b=[5], d=[9])

In [3]: Batch.cat([a,b,c])
Out[3]: 
Batch(
    b: array([2, 5, 5]),
    a: array([1, 4, 0]),
    c: array([3, 0, 3]),
    d: array([0, 6, 9]),
)

In [4]: Batch.stack([a,b,c])
Out[4]: 
Batch(
    b: array([[2],
              [5],
              [5]]),
)
```

Reply:
It is not related to this part of the code, otherwise this would not work
```
In [1]: from tianshou.data import Batch                                                                                                                                                                     

In [2]: a = Batch(a=[1], b=[2], c=[3])                                                                                                                                                                      

In [3]: b = Batch(a=[4], b=[5], d=[6])                                                                                                                                                                      

In [4]: Batch.stack([a,b])                                                                                                                                                                                  
Out[4]: 
Batch(
    a: array([[1],
              [4]]),
    b: array([[2],
              [5]]),
    d: array([[0],
              [6]]),
    c: array([[3],
              [0]]),
)
```

Reply:
You can check by yourself that simply adding your modification line 604 `set.union(*keys_map) - keys_shared` does the trick:
```
In [1]: from tianshou.data import Batch                                                                                                                                                                     

In [2]: a = Batch(a=[1], b=[2], c=[3])                                                                                                                                                                      

In [3]: b = Batch(a=[4], b=[5], d=[6])                                                                                                                                                                      

In [4]: c = Batch(c=[3], b=[5], d=[9])                                                                                                                                                                      

In [5]: Batch.stack([a,b,c])                                                                                                                                                                                
Out[5]: 
Batch(
    b: array([[2],
              [5],
              [5]]),
    c: array([[3],
              [0],
              [3]]),
    a: array([[1],
              [4],
              [0]]),
    d: array([[0],
              [6],
              [9]]),
)
```

Reply:
Batch.stack([a,b]) work but Batch.stack([a,b,c]) does not work.

Reply:
Yes it is, see my comment just above :)

Reply:
Try three Batches, the desired behavior is to remain all keys, but previous code only reserves some keys.

Reply:
> Try three Batches, the desired behavior is to remain all keys, but previous code only reserves some keys.

How many times I need to tell you to read my comment ?

> You can check by yourself that simply adding your modification line 604 set.union(*keys_map) - keys_shared does the trick

Reply:
Also, I strongly suggest to simplify your `cat` implementation to follow the same pattern than the current implementation of `cat`, that is clearer and easy to maintain in my opinion.

Reply:
In the first place, I try to re-implement ``Batch.cat_``, and the "duplicate" code is required (#128 ). Then I remember to re-implement ``Batch.stack`` and copied the code. I'll check later if they are necessary, or you can help me check it :)

Reply:
> Also, I strongly suggest to simplify your `cat` implementation to follow the same pattern than the current implementation of `cat`, that is clearer and easy to maintain in my opinion.

Strongly AGREE!!!

Reply:
> Also, I strongly suggest to simplify your `cat` implementation to follow the same pattern than the current implementation of `cat`, that is clearer and easy to maintain in my opinion.

I'll try. Or you can help me. ``cat`` is more complicated than ``stack``. The shapes of stack arguments are the same, but they differ in cat.

Reply:
The way you are dealing all this changes is impossible to follow and I'm thinking to giving up.

You should focus more on how to make existing things works rather than reimplementing stuff. I know its may be frustrating but that's the only way to collaborate with people. Keep in mind other people are not 100% involve in your project.

So it is up to you, but more than 2 PR (<500 lines) a day is clearer impossible to follow.

Reply:
The reason I want to re-implement is because I do not understand what ``keys_partial = reduce(set.symmetric_difference, keys_map)`` does when having more than two sets. I thought that you depend on some magic that I do not know, so I re-implement it  =_=

Reply:
Emm, I'm sorry for this.

Reply:
> Emm, I'm sorry for this.

Don't be. You did nothing wrong but your best I'm pretty sure. Plus, replacing `keys_partial = reduce(set.symmetric_difference, keys_map)` by `keys_partial = set.union(*keys_map) - keys_shared` is VERY GOOD and fixes the bug with `stack` !

I'm not saying that you are doing anything wrong. Just try to keep in mind that open-source is great because it involves a community. So it means several things: many people are happy to spend times to improve the project, BUT, you also need to compromise refrain your enthusiasm of changing many stuffs at once. Other people can only follow small steps. Trust me I know it is frustrating to spend time trying to fix code developed by others and is time consuming, but it is the only way to get and keep other people involved.

Reply:
Thank you. I have learned so much from you and from this project. ♡

I'll fix them tomorrow :)

Reply:
Last but not last, you are doing a very good job so far, it just seems you need to learn to step back a little bit and get the full picture.

Reply:
> Thank you. I have learned so much from you and from this project. ♡
I'll fix them tomorrow :)

And it is just the beginning!